### PR TITLE
Rebuild with OpenSSL 3 [skip ci]

### PR DESCRIPTION
**Pull request by JeanChristopheMorinPerso**
Rebuild with OpenSSL 3.

Note that openssl was missing in the host dependencies and also on osx-64 and Linux! On windows, it uses the Windows APIs instead of OpenSSL.

### CODE REVIEW

**Review by skupr-anaconda — APPROVED** [review]
LGTM

**Comment by danpetry** [comment]
Does it use openssl automatically on mac? See [here](https://github.com/Unidata/netcdf-c/blob/3e1280b24bd5dfc899b841b38e1bc0e0eac204b2/libdispatch/nccurl_setup.h#L38)

**Comment by danpetry** [comment]
Also, do you know, why was it a missing dependency in the first place? Did it simply not build and run, or is it an optional dependency..?

File: recipe/meta.yaml [hunk 1]
```diff
@@ -62,55 +61,29 @@ requirements:
...
-    - jpeg
-    - libzip
+    - bzip2 1.0.8
+    - libcurl 7.88.1
+    - zlib {{ zlib }}
+    - hdf4 4.2.13
+    - hdf5 1.12.1
+    - jpeg 9e
+    - libzip 1.8.0
+    - openssl {{ openssl }}  # [not win]
```
Comment (boldorider4) [inline recipe/meta.yaml:71]: Maybe to me this is still unclear, but why is the pinning inherited from `cbc.yaml` in the host section if this has to be built agains `openssl 3.*`?

Comment (danpetry) [reply recipe/meta.yaml:71]: it's using an aggregate branch called openssl3_builds (see the abs.yaml). In this branch, the aggregate cbc.yaml has the openssl 3 pinning

Comment (boldorider4) [reply recipe/meta.yaml:71]: gotcha. So I guess it's ok to keep this with the knowledge that aggregate will soon be update, right?

Comment (danpetry) [reply recipe/meta.yaml:71]: Well it's kind of a question of the tradeoffs/approach that @JeanChristopheMorinPerso has chosen. Also related to the discussion [here](https://github.com/AnacondaRecipes/cyrus-sasl-feedstock/pull/5)

Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml:71]: Yeah. Technically it's a really bad practice, but unfortunately it's a practice that the distro team has been doing for a while. We've decided to keep doing it for now and discuss it later. There is already some talks in progress regarding how we pin packages, for example to get rid of implicit dependencies, etc but it'll take while I think before we can reach a consensus.

**Comment by JeanChristopheMorinPerso** [comment]
>Does it use openssl automatically on mac? See [here](https://github.com/Unidata/netcdf-c/blob/3e1280b24bd5dfc899b841b38e1bc0e0eac204b2/libdispatch/nccurl_setup.h#L38)

@danpetry I think it's used on macOS... Conda build says
```
   INFO: sysroot: '/Library/Developer/CommandLineTools/SDKs/MacOSX11.1.sdk/' files: '['usr/share/man/mann/zero.n', 'usr/share/man/mann/yencode.n', 'usr/share/man/mann/yaml.n', 'usr/share/man/mann/xsxp.n']'
   INFO (libnetcdf,bin/ncdump): Needed DSO lib/libnetcdf.19.dylib found in the whitelist
   INFO (libnetcdf,bin/ncdump): Needed DSO lib/libmfhdf.0.dylib found in pkgs/main::hdf4-4.2.13-h5e329fb_3
   INFO (libnetcdf,bin/ncdump): Needed DSO lib/libdf.0.dylib found in pkgs/main::hdf4-4.2.13-h5e329fb_3
   INFO (libnetcdf,bin/ncdump): Needed DSO lib/libhdf5_hl.200.dylib found in staging_openssl3/label/openssl3::hdf5-1.12.1-h05c076b_3
   INFO (libnetcdf,bin/ncdump): Needed DSO lib/libhdf5.200.dylib found in staging_openssl3/label/openssl3::hdf5-1.12.1-h05c076b_3
   INFO (libnetcdf,bin/ncdump): Needed DSO lib/libcrypto.3.dylib found in staging_openssl3/label/openssl3::openssl-3.0.8-h1a28f6b_0
   INFO (libnetcdf,bin/ncdump): Needed DSO lib/libcurl.4.dylib found in staging_openssl3/label/openssl3::libcurl-7.88.1-h3e2b118_1
```

Unless macOS provides a libcrypto too?

>Also, do you know, why was it a missing dependency in the first place? Did it simply not build and run, or is it an optional dependency..?

The first thing was that I saw it was in the run dependencies but not int he host dependency. That was the first red flag. Then I looked at the code and saw it is using openssl. And I also looked at the linker flags. Lastly, I think conda-build was complaining.

File: recipe/meta.yaml (hunk 1 above)
Comment (danpetry) [reply recipe/meta.yaml:71]: As far as i understand, Jean-Christophe isn't going to merge to master until the aggregate is updated

Comment (danpetry) [reply recipe/meta.yaml:71]: @JeanChristopheMorinPerso just spoke to Rafael, he said that you are doing a mutual staging branch thing between you and the ICU rebuild, then when the packages are built out you'll merge the branches containing the updated aggregate pinnings for openssl and icu, then merge the staging branches? So then, the aggregate cbc.yaml pinning will be updated before the updated feedstocks are merged.

I'm also wondering why we wouldn't just merge the updated aggregate/cbc.yaml first, then rebuild and merge the packages one by one?

Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml:71]: >he said that you are doing a mutual staging branch thing between you and the ICU rebuild

Umm, I suggested that we could collaborate on Qt to avoid having to rebuild Qt multiple times (since Qt is an heavy and painful package).

@ryanskeith can add more details, but we wanted to be extra careful. We didn't want to force OpenSSL 3 on everyone too early. We are planning on merging the aggregate branch in master on Monday.

Comment (danpetry) [reply recipe/meta.yaml:71]: that makes sense intuitively, I'm just wondering if it really makes a difference, since everything that depends on openssl is being rebuilt anyway. And then if someone updates one of those packages before you can get to it, they've done the work for you?

**Comment by JeanChristopheMorinPerso** [comment]
@danpetry I think I understand what's going on with OpenSSL. The netcdf libraries and executables link against libcrypto on Mac and Linux. The code you pointed to is not for our version though. In reality, OpenSSL isn't used directly by netcdf, at least for 4.8.1. It will in a future non-yet-released version (the code you pointed at). But because of the way to link against hdf5, libcrypto gets pulled in and is linked against netcdf libraries.

They use `find_package` to get hdf5 (https://github.com/Unidata/netcdf-c/blob/v4.8.1/CMakeLists.txt#L652) and then they use `HDF5_C_LIBRARIES` to get the actual HDF5 library (https://github.com/Unidata/netcdf-c/blob/v4.8.1/liblib/CMakeLists.txt#L93). CMake has native support for hdf5 (https://cmake.org/cmake/help/latest/module/FindHDF5.html#result-variables). As per the documentation, `HDF5_C_LIBRARIES` represents the required libraries for the HDF5 C bindings.

If I print the content of that variable, I get
```
/Users/jcmorin/miniconda/conda-bld/debug_1685737493665/_h_env/lib/libhdf5.dylib
/Users/jcmorin/miniconda/conda-bld/debug_1685737493665/_h_env/lib/libcrypto.dylib
/Users/jcmorin/miniconda/conda-bld/debug_1685737493665/_h_env/lib/libcurl.dylib
/Library/Developer/CommandLineTools/SDKs/MacOSX11.1.sdk/usr/lib/libpthread.tbd
/Users/jcmorin/miniconda/conda-bld/debug_1685737493665/_h_env/lib/libz.dylib
/Library/Developer/CommandLineTools/SDKs/MacOSX11.1.sdk/usr/lib/libdl.tbd
/Library/Developer/CommandLineTools/SDKs/MacOSX11.1.sdk/usr/lib/libm.tbd
```

So basically, their CMake files are kind of bad... They should use the `hdf5::hdf5` target instead of `HDF5_C_LIBRARIES` to avoid linking against transitive libraries.

**Comment by JeanChristopheMorinPerso** [comment]
There is one thing I'm not sure about though. On Windows, conda-build complains and says
```
INFO   WARNING (libnetcdf): run-exports library package staging_openssl3/label/openssl3::openssl-3.0.8-h2bbff1b_0 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to   
 `build/ignore_run_exports`)    
```

I'm not sure why the run_exports of openssl is being applied.

**Comment by danpetry** [comment]
Maybe it's being pulled in as a transitive dependency?

**Comment by JeanChristopheMorinPerso** [comment]
If `run_exports` are applied for transitive dependencies, `run_exports` is much worst than what I thought (and I already thought it was quite bad in the first place)...

**Comment by danpetry** [comment]
yeah true. they are not

**Comment by danpetry** [comment]
`libcurl` in `host` is pulling it in for some reason

**Comment by JeanChristopheMorinPerso** [comment]
I think I found the issue. I made a mistake in hdf5 and added `openssl` in the `run_exports` section instead of the `ignore_run_exports`, see https://github.com/AnacondaRecipes/hdf5-feedstock/pull/12/files#diff-f3725a55bf339595bf865fec73bda8ac99f283b0810c205442021f29c06eea9aR34-R35. I pushed a new commit in hdf5, and it should fix the issue. So I've re-triggered builds for libnetcdf and will confirm once the builds are finished that it's fixed.

**Comment by JeanChristopheMorinPerso** [comment]
Yep, problem fixed:

`conda search libnetcdf -i -c staging_openssl3/label/openssl3 --override-channels --subdir osx-arm64`:
```
dependencies: 
  - curl
  - hdf4 >=4.2.13,<4.2.14.0a0
  - hdf5 >=1.12.1,<1.12.2.0a0
  - libcurl >=7.88.1,<8.0a0
  - libcxx >=14.0.6
  - libzip
  - openssl 3.*
  - openssl >=3.0.8,<4.0a0
  - zlib
```
`conda search libnetcdf -i -c staging_openssl3/label/openssl3 --override-channels --subdir win-64`:
```
dependencies: 
  - curl
  - hdf4 >=4.2.13,<4.2.14.0a0
  - hdf5 >=1.12.1,<1.12.2.0a0
  - libcurl >=7.88.1,<8.0a0
  - libzip
  - vc >=14.1,<15.0a0
  - vs2015_runtime >=14.16.27012,<15.0a0
  - zlib
```

**Review by danpetry — APPROVED** [review]
